### ISSUE TITLE
Field becomes dirty when using dynamic currency-symbol

### ISSUE THREAD

**Issue by houmark** [aguirrel/ng-currency#102]
### Description

We use angular-locale to set the formatting of currencies etc. I just updated from a pre 0.9 version and noticed the field/form is now dirty after loading the form and since we have a dirty tracker that alerts the user when the form is "unsaved" this alert is now persistent.

### Expected outcome

That the field is not becoming dirty from the currency symbol change.

### Actual outcome

Field becomes dirty.

### Live Demo

The below demo is a modified version of the main example. I changed local to Danish so it will load `kr` by default and then in the scope I change it to `USD`. It's very basic to re-create. If you keep your eyes sharp, you will notice the initial `kr` changing to `USD` and also that the form is first `dirty: false` and then becomes `dirty: true`
http://plnkr.co/edit/oMsst0D0igUst6TOZ6TV?p=preview

### Steps to reproduce

See demo (and the simple code changes). Removing `currency-symbol` from the HTML stops the form from becoming dirty but of course it makes it impossible to use a dynamic currency symbol.

### Browsers Affected

- [x] Chrome
- [x] Firefox
- [x] Safari 9
- [ ] Safari 8 (did not test)
- [ ] Safari 7 (did not test)
- [ ] Edge (did not test)
- [ ] IE 11 (did not test)
- [ ] IE 10 (did not test)

### Off topic:

I kinda like the new way of hiding the currency when the user enters the field, but it seems the model value needs to used for that and not just the view value, which then changes the model and that means the field becomes dirty unless you use the default currency symbol. A hacky way may be to $setPristine() right after loading the symbol, which should be safe enough since the user won't be able to enter the field before, but in theory other code could, which would reset the dirty status, and that could raise the opposite problem/situation.

I also don't like much that the field changes to `22.11` when focusing, as I as would expect it to stay `22,11` which is the Danish formatting (and how it worked pre 0.9.x). I do understand that I can type in `22,12`, but it's confusing for the user to see it change from the correct formatting to what would be considered thousands when editing (by people in countries that uses comma for decimals - and dots for thousands).

In any case, thank you for a great directive!

**Comment by aguirrel** [comment]
Hi @houmark! We are checking $dirty problem.

About changing format on focus, I'm in Argentina and we use comma (,) for decimal separator too, but to be honest, all keyboards use dot (.) in numeric keypad and I think is a mess to use numeric keypad and change to alphanumeric keypad only to introduce the comma, also many mobiles browsers use dot(.) when you define fields like number. :)

Best regards,

Luis

**Comment by houmark** [comment]
I understand your point, but from a usability perspective and to prevent misunderstandings with thousands vs. decimals, it should use the comma for decimals. Could it not be configurable somehow? Or auto loaded from the locale? I know many other things has to be taken into consideration, but if the $dirty issue is fixed and the right decimal marker is shown along with the currency moving out of the way, this is the perfect combo imho :)

Hoping you can find a solution to the $dirty issue. Thanks for your fast reply.

**Comment by aguirrel** [comment]
I agree with the decimal marker, could you add a new issue with this so we don't mix this two differents things? :) Thaks a lot!

**Comment by houmark** [comment]
Sure will :)

**Comment by nick-woodward** [comment]
My guess is that our modifications of the $viewValue outside of parsers/formatters is causing the input to become dirty.

I need to verify, but I'm guessing if we check the pristine value prior to currencyRedraw running it returns true, whereas after the currencyRedraw it becomes dirty.

This also may be true for on focus since it also modifies the $viewValue.

**Comment by houmark** [comment]
I forked with the intent to try a fix but I must admit that there's a bit too much going on here for me to understand all logic. I've read that `ngModel.$render();` right after `ngModel.$setViewValue(viewValue);` is a possible fix but it looks like that's already done more or less all places.

Another solution is not binding to ngModel and then keep a copy of the model value internally and only update the model once the value changes. See more [here](http://stackoverflow.com/questions/22878064/angularjs-set-view-value-is-not-updating-display), [here](http://stackoverflow.com/questions/30538717/ngmodel-how-to-manipulate-initial-model-value-without-making-it-dirty) (setPristine hack if was pristine before) or [here](http://stackoverflow.com/questions/30538717/ngmodel-how-to-manipulate-initial-model-value-without-making-it-dirty) which seems to be the most correct way of solving it (see comments also which explains the details).

So for now I'm back to using 0.8.x in order to keep my app working...

**Comment by aguirrel** [comment]
Not solved all use cases, reopening it.

**Comment by aguirrel** [comment]
Hi guys,

I did something like [this](http://stackoverflow.com/questions/30538717/ngmodel-how-to-manipulate-initial-model-value-without-making-it-dirty)

New code:

```
      $timeout(() => {
        initialized = true;
        ngModel.$pristine = false;
        refresh();
        ngModel.$pristine = true;
      });
```
It's a smelly hack, you have to see $setViewValue inside angular to understand why $pristine needs to be false. 

@nick-woodward if you have any other idea is very welcome. :smile: 

@houmark please update to version 0.11.3 to get this fixed.

Best regards.

**Comment by houmark** [comment]
Thanks for your work and attempts. Unfortunately your fix does not work for me.

I'm not sure `$pristine = true` is a safe way of getting things pristine. Since that's a boolean, it will simply "hack" the current field to be considered pristine, but the entire form won't really know. `$setPristine()` should work better for that. Having said all of that, all the fields in my form using ng-currency is still dirty with latest master.

It may have something to do with how our forms are loaded and the data, but I think ng-currency should not be affected by that if things are done right. Did you consider the method of copying the ngModel value on load, using that and only updating the model if the data changes (the last link I included in my last comment)?

**Comment by houmark** [comment]
With #46a8fe4 I do get the field pristine, but not the form, which is weird, since `$setPristine()` on a field should also hook in with the form. I will test more, but I have a feeling a rewrite of the concept is needed to make this truly stable.

**Comment by houmark** [comment]
After more testing with latest master (I suspect I did not get the code in with master the last time I tested) the field does become pristine, but since the form had become dirty from the field being dirty — even if only shortly — the form stays dirty in my setup which loads data from our API, so while it seems like the fix is working in the plunker, in a more real life situation it does not, do to what could be considered a race situation.

Another side effect is also that clicking into the field (focusing) sets it dirty even if you do not make any change to the value, so my best suggestion would be to try to isolate the value on load and avoid modifying the model value until the value actually changes - and then use $setPristine() when changing the view value. Again, I would like to attempt to fix it myself, but it's a bit hard for me to go in a do "major" changes here without risking that something else breaks, and the directive does quite a lot of stuff, I might not spot issues introduced by my changes.
